craft_parts/executor/part_handler.py

```diff
@@ -848,18 +901,23 @@ def _clean_shared(self, step: Step, *, shared_dir: Path) -> None:
             overlay_migration_state=overlay_migration_state,
         )
 
+        parts_with_overlay_in_step = _parts_with_overlay_in_step(
+            step, part_list=self._part_list
+        )
+        logger.info(f"parts_with_overlay_in_step: {parts_with_overlay_in_step}")
+
         # remove overlay data if this is the last part with overlay
-        if (
-            self._part.has_overlay
-            and len(_parts_with_overlay_in_step(step, part_list=self._part_list)) == 1
-        ):
+        if self._part.has_overlay and len(parts_with_overlay_in_step) == 1:
             migration.clean_shared_overlay(
                 shared_dir=shared_dir,
                 part_states=part_states,
                 overlay_migration_state=overlay_migration_state,
             )
             overlay_migration_state_path = states.get_overlay_migration_state_path(
-                self._part.overlay_dir, step
+                self._part.overlay_dirs[partition], step
+            )
+            logger.info(
+                f"remove overlay migration state file for part {self._part.name}, step {step}"
```

Comment:
I am not sure we want to keep this. That was useful to debug but might not really needed in production.